syntax = "proto3";

package tetra;

// The remote execution service definition
service RemoteExecutor {
  // Execute a function remotely
  rpc ExecuteFunction (FunctionRequest) returns (FunctionResponse) {}
}

// The request message containing function details and arguments
message FunctionRequest {
  string function_name = 1;       // Name of the function to execute
  string function_code = 2;       // Source code of the function
  repeated string args = 3;       // Base64-encoded cloudpickle-serialized positional arguments
  map<string, string> kwargs = 4; // Base64-encoded cloudpickle-serialized keyword arguments
  repeated string dependencies = 5; // Optional list of pip packages to install before execution
}

// The response message containing the execution result or error
message FunctionResponse {
  bool success = 1;               // Whether execution was successful
  string result = 2;              // Base64-encoded cloudpickle-serialized result (if success)
  string error = 3;               // Error message (if not success)
}